pygeos/_bulk_creation.pyx

```diff
@@ -0,0 +1,144 @@
+    if indices.ndim != 1:
+        raise TypeError("indices is not a one-dimensional array.")
+
+    if geometries.shape[0] != indices.shape[0]:
+        raise ValueError("geometries and indices do not have equal size.")
+
+    if geometries.shape[0] == 0:
+        # return immediately if there are no geometries to return
+        return np.empty(shape=(0, ), dtype=np.object_)
+
+    if np.any(indices[1:] < indices[:-1]):
+        raise ValueError("The indices should be sorted.")  
+
+    cdef object[:] geometries_view = geometries
+    cdef int[:] indices_view = indices
+
+    # get the geometry count per collection
+    cdef long[:] collection_size = np.bincount(indices)
+
+    # A temporary array for the geometries that will be given to CreateCollection.
+    # Its size equals max(collection_size) to accomodate the largest collection.
+    temp_geoms = np.empty(shape=(np.max(collection_size), ), dtype=np.intp)
+    cdef np.intp_t[:] temp_geoms_view = temp_geoms
+
+    # The final target array
+    cdef Py_ssize_t n_colls = collection_size.shape[0]
+    result = np.empty(shape=(n_colls, ), dtype=np.object_)
+    cdef object[:] result_view = result
+
+    with get_geos_handle() as geos_handle:
+        for coll_idx in range(n_colls):
+            coll_size = collection_size[coll_idx]
+
+            # fill the temporary array with geometries belonging to this collection
+            for this_geom_idx in range(coll_size):
+                geom_idx = first_geom_idx + this_geom_idx
+                if PyGEOS_GetGEOSGeometry(<PyObject *>geometries_view[geom_idx], &geom) == 0:
+                    # deallocate previous temp geometries (preventing memory leaks)
+                    for geom_idx in range(this_geom_idx):
+                        GEOSGeom_destroy_r(geos_handle, <GEOSGeometry *>temp_geoms_view[geom_idx])
+                    raise TypeError(
+                        "One of the arguments is of incorrect type. Please provide only Geometry objects."
+                    )
+
+                # Check geometry subtype for non-geometrycollections
+                if geometry_type != 7:
+                    curr_type = GEOSGeomTypeId_r(geos_handle, geom)
+                    if curr_type != expected_type and curr_type != expected_type_alt:
+                        # deallocate previous temp geometries (preventing memory leaks)
+                        for geom_idx in range(this_geom_idx):
+                            GEOSGeom_destroy_r(geos_handle, <GEOSGeometry *>temp_geoms_view[geom_idx])
+                        raise TypeError(
+                            f"One of the arguments has unexpected geometry type {curr_type}."
+                        )
+
+                # ignore missing values
+                if geom == NULL:
+                    n_missing += 1
+                else:
+                    # assign to the temporary geometry array
+                    temp_geoms_view[this_geom_idx] = <np.intp_t>GEOSGeom_clone_r(geos_handle, geom)
```

Comment:
`this_geom_idx` is the position within the inner loop, but shouldn't be used for saving this into the array because it doesn't account for missing geometries.

Instead, I'd suggest using another counter that is incremented as you set non-empty geometries into `temp_geoms_view` and then use that value as the length of the geometry collection for `GEOSGeom_createCollection_r` below.


Here is a test case to add to `test_collections_1d` for that (currently segfaults):
```python
([point, None, line_string], [0, 0, 0], [geom_coll([point, line_string])])
```